src/Types/RelationParserHelper.php

```diff
@@ -7,24 +7,46 @@
 use PhpParser\Node;
 use PhpParser\Node\Expr\MethodCall;
 use PhpParser\NodeFinder;
+use PHPStan\Analyser\ScopeContext;
+use PHPStan\Analyser\ScopeFactory;
 use PHPStan\Parser\CachedParser;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\Generic\GenericClassStringType;
+use PHPStan\Type\Generic\TemplateTypeMap;
+use PHPStan\Type\ObjectType;
 
 class RelationParserHelper
 {
     /** @var CachedParser */
     private $parser;
 
-    public function __construct(CachedParser $parser)
+    /** @var ScopeFactory */
+    private $scopeFactory;
+
+    public function __construct(CachedParser $parser, ScopeFactory $scopeFactory)
     {
         $this->parser = $parser;
+        $this->scopeFactory = $scopeFactory;
     }
 
-    public function findRelatedModelInRelationMethod(string $fileName, string $methodName): ?string
-    {
+    public function findRelatedModelInRelationMethod(
+        MethodReflection $methodReflection
+    ): ?string {
+        $fileName = $methodReflection
+            ->getDeclaringClass()
+            ->getNativeReflection()
+            ->getMethod($methodReflection->getName())
+            ->getFileName();
```

Comment:
If the method is defined in a trait `$methodReflection->getDeclaringClass()->getFileName()` wouldn't return the trait filename. So this is the only way I found to get the trait filename if it's defined in a trait. @ondrejmirtes Is there a better way?

Reply:
I don't think this will always return the trait 🤔

There's no easy way in PHP reflection to find out if a method comes from a trait or not. PHPStan itself does some complex logic to find out about it (https://github.com/phpstan/phpstan-src/blob/master/src/Reflection/Php/PhpClassReflectionExtension.php#L640-L674) to get the right PHPDocs but even this sometimes fails.

There's a nice article about it: http://mouf-php.com/blog/php_reflection_api_traits

What I'd recommend to you is to have guards around this code - if you don't find the method definition in the filename that reflection reports (so it means it's somewhere else, like in a trait), just return `null` or `mixed` or whatever you need to return to not analyse that piece of code..

Reply:
Oh, ok. Good to know :+1: 

Yeah, there are already some guards. If `getFileName` returns false, or if we can't find the method we are looking for in the file, it'll not analyze it further.